[generic-worker] Change ownership of caches to current task

**Describe the bug**
As of Git 2.35.2, they introduced a security feature to mitigate a CVE on multi-user setups:
https://github.blog/2022-04-12-git-security-vulnerability-announced/#cve-2022-24765

Basically Git will error out if the checkout dir isn't owned by the current user. This impacts Firefox-CI because we commonly cache the checkout dir of the Github repo. So it will be owned by the first task user to check it out, and then the second task user will fail as it does not own the checkout.

**To Reproduce**

Here's an example task that hit this issue:
https://firefox-ci-tc.services.mozilla.com/tasks/cVeBuKGhQkGDqzsoz_utkA/runs/0/logs/public/logs/live.log

Relevant log:
```
[taskcluster 2023-09-18T18:20:35.232Z] Executing command 0: C:/mozilla-build/python3/python3.exe run-task --mozillavpn-checkout=./build/src/ --task-cwd build/src/windows/installer -- cmd.exe /c build_prod.cmd
D:\task_169506064099579>C:/mozilla-build/python3/python3.exe run-task --mozillavpn-checkout=./build/src/ --task-cwd build/src/windows/installer -- cmd.exe /c build_prod.cmd 
[setup 2023-09-18T18:20:35.398Z] run-task started in D:\task_169506064099579
[setup 2023-09-18T18:20:35.398Z] Invoked by command: --mozillavpn-checkout=./build/src/ --task-cwd build/src/windows/installer -- cmd.exe /c build_prod.cmd
[setup 2023-09-18T18:20:35.398Z] Python version: 3.9.10
[vcs 2023-09-18T18:20:35.401Z] WARNING: vcs checkout path (D:\task_169506064099579\build\src) not in cache or volume; performance will likely suffer
[vcs 2023-09-18T18:20:35.401Z] executing ['git', 'fetch', '--tags', '--force', 'https://github.com/mozilla-mobile/mozilla-vpn-client', 'refs/heads/main']
[vcs 2023-09-18T18:20:35.414Z] fatal: detected dubious ownership in repository at 'D:/task_169506064099579/build/src'
[vcs 2023-09-18T18:20:35.414Z] 'D:/task_169506064099579/build/src' is owned by:
[vcs 2023-09-18T18:20:35.414Z] 	'S-1-5-21-867665065-978626105-2054798079-1000'
[vcs 2023-09-18T18:20:35.414Z] but the current user is:
[vcs 2023-09-18T18:20:35.414Z] 	'S-1-5-21-867665065-978626105-2054798079-1001'
[vcs 2023-09-18T18:20:35.414Z] To add an exception for this directory, call:
[vcs 2023-09-18T18:20:35.414Z] 
[vcs 2023-09-18T18:20:35.414Z] 	git config --global --add safe.directory D:/task_169506064099579/build/src
```

**Expected behavior**

We worked around this by disabling the `safe.directory` feature as the error recommends:
https://github.com/taskcluster/taskgraph/pull/346

However the safety check seems like it could actually prove useful. I think `generic-worker` should handle updating cache ownership to the current user which would avoid this issue. Though workers that run more than one task at a time might prove to be a challenge :/

We're hitting real world issues with this now in taskgraph. Specifically, tasks like https://firefox-ci-tc.services.mozilla.com/tasks/OWIQjgQmSpuwEbCyweZdwg are running tests that require the git directory to be owned by the current task user. With the current state of things, this results in failures such as:
```
[task 2026-05-19T21:21:55.138+00:00] During handling of the above exception, another exception occurred:
[task 2026-05-19T21:21:55.138+00:00] 
[task 2026-05-19T21:21:55.138+00:00] mock_files_changed = <MagicMock name='get_changed_files' id='2495164596304'>
[task 2026-05-19T21:21:55.138+00:00] mock_get_note = <MagicMock name='get_note' id='2495171338896'>
[task 2026-05-19T21:21:55.138+00:00] 
[task 2026-05-19T21:21:55.138+00:00]     @unittest.mock.patch.object(
[task 2026-05-19T21:21:55.138+00:00]         GitRepository,
[task 2026-05-19T21:21:55.138+00:00]         "get_note",
[task 2026-05-19T21:21:55.138+00:00]         return_value="not valid json {",
[task 2026-05-19T21:21:55.138+00:00]     )
[task 2026-05-19T21:21:55.138+00:00]     @unittest.mock.patch.object(GitRepository, "get_changed_files", return_value=[])
[task 2026-05-19T21:21:55.138+00:00]     def test_decision_parameters_note_invalid_json(mock_files_changed, mock_get_note):
[task 2026-05-19T21:21:55.138+00:00]         options = {**_BASE_OPTIONS, "allow_parameter_override": True}
[task 2026-05-19T21:21:55.138+00:00] >       with pytest.raises(
[task 2026-05-19T21:21:55.138+00:00]             Exception, match="Failed to parse refs/notes/decision-parameters as JSON"
[task 2026-05-19T21:21:55.138+00:00]         ):
[task 2026-05-19T21:21:55.138+00:00] E       AssertionError: Regex pattern did not match.
[task 2026-05-19T21:21:55.138+00:00] E         Expected regex: 'Failed to parse refs/notes/decision-parameters as JSON'
[task 2026-05-19T21:21:55.138+00:00] E         Actual message: "Command '('C:\\\\Program Files\\\\Git\\\\cmd\\\\git.EXE', 'log', '-n1', '--format=%B', 'HEAD')' returned non-zero exit status 128."
[task 2026-05-19T21:21:55.138+00:00] 
[task 2026-05-19T21:21:55.138+00:00] test\test_decision.py:197: AssertionError
[task 2026-05-19T21:21:55.138+00:00] ---------------------------- Captured stderr call -----------------------------
[task 2026-05-19T21:21:55.138+00:00] fatal: detected dubious ownership in repository at 'D:/task_177922473516860/build/src'
[task 2026-05-19T21:21:55.138+00:00] 'D:/task_177922473516860/build/src' is owned by:
[task 2026-05-19T21:21:55.138+00:00] 	(inconvertible) (S-1-5-21-3609005065-3965640373-874575067-1004)
[task 2026-05-19T21:21:55.138+00:00] but the current user is:
[task 2026-05-19T21:21:55.138+00:00] 	VM-DPMWKWMT3MTV/task_177922473516860 (S-1-5-21-3609005065-3965640373-874575067-1006)
[task 2026-05-19T21:21:55.138+00:00] To add an exception for this directory, call:
[task 2026-05-19T21:21:55.138+00:00] 
[task 2026-05-19T21:21:55.138+00:00] 	git config --global --add safe.directory D:/task_177922473516860/build/src
[task 2026-05-19T21:21:55.138+00:00] =========================== short test summary info ===========================
```

We can work around this the same way, but with Firefox moving quickly towards running tasks against Git repos I suspect more of this will come up, and that we ought to fix this sooner rather than later.

--- a/changelog/issue-6561.md
+++ b/changelog/issue-6561.md
@@ -1,0 +1,6 @@
+audience: worker-deployers
+level: patch
+reference: issue 6561
+---
+On Windows, generic-worker now transfers ownership of mounts in addition to
+giving the task user full control.

--- a/workers/generic-worker/mounts_multiuser_windows_test.go
+++ b/workers/generic-worker/mounts_multiuser_windows_test.go
@@ -1,0 +1,85 @@
+//go:build multiuser
+
+package main
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/mcuadros/go-defaults"
+	"github.com/taskcluster/taskcluster/v100/workers/generic-worker/host"
+)
+
+func ownerOf(t *testing.T, path string) string {
+	t.Helper()
+	out, err := host.Output("powershell", "-NoProfile", "-NonInteractive", "-Command",
+		"(Get-Acl -LiteralPath '"+path+"').Owner")
+	if err != nil {
+		t.Fatalf("could not read owner of %q: %v", path, err)
+	}
+	owner := strings.TrimSpace(out)
+	return owner[strings.LastIndex(owner, `\`)+1:]
+}
+
+func TestWritableDirectoryCacheReclaimsOwnership(t *testing.T) {
+	setup(t)
+
+	mounts := []MountEntry{
+		&WritableDirectoryCache{
+			CacheName: "banana-cache",
+			Directory: "ownership-cache",
+		},
+	}
+
+	payload := GenericWorkerPayload{
+		Mounts:     toMountArray(t, &mounts),
+		Command:    helloGoodbye(),
+		MaxRunTime: 180,
+	}
+	defaults.SetDefaults(&payload)
+
+	runTask := func() {
+		td := testTask(t)
+		td.Scopes = append(td.Scopes, "generic-worker:cache:banana-cache")
+		_ = submitAndAssert(t, td, payload, "completed", "completed")
+	}
+
+	cacheLocation := func() string {
+		entries := directoryCaches["banana-cache"]
+		if len(entries) == 0 {
+			t.Fatal("expected a persisted banana-cache entry after the task ran")
+		}
+		return entries[0].Location
+	}
+
+	runTask()
+	location := cacheLocation()
+
+	nested := filepath.Join(location, "sub", "file.txt")
+	if err := os.MkdirAll(filepath.Dir(nested), 0700); err != nil {
+		t.Fatal(err)
+	}
+	if err := os.WriteFile(nested, []byte("data"), 0600); err != nil {
+		t.Fatal(err)
+	}
+
+	// *S-1-5-32-544 is the BUILTIN\Administrators group
+	if err := host.Run("icacls", location, "/setowner", "*S-1-5-32-544", "/T"); err != nil {
+		t.Fatalf("could not reassign cache ownership: %v", err)
+	}
+	if owner := ownerOf(t, location); owner == taskContext.User.Name {
+		t.Fatalf("cache %q is still owned by task user %q before reuse", location, taskContext.User.Name)
+	}
+
+	runTask()
+	location = cacheLocation()
+	nested = filepath.Join(location, "sub", "file.txt")
+
+	for _, p := range []string{location, nested} {
+		if owner := ownerOf(t, p); owner != taskContext.User.Name {
+			t.Errorf("expected %q to be owned by task user %q after cache reuse, but owner is %q", p, taskContext.User.Name, owner)
+		}
+	}
+}

--- a/workers/generic-worker/multiuser_windows.go
+++ b/workers/generic-worker/multiuser_windows.go
@@ -267,8 +267,20 @@ func install(arguments map[string]any) (err error) {
 }
 
 func makeFileOrDirReadWritableForUser(recurse bool, dir string, user *gwruntime.OSUser) error {
-	// see http://ss64.com/nt/icacls.html
-	return host.Run("icacls", dir, "/grant:r", user.Name+":(OI)(CI)F")
+	// On windows, grant:r or ownership aren't enough on their own. Ownership
+	// doesn't mean it's going to be readable, and ownership is necessary for
+	// programs like git that check that files haven't been tampered with (see #6561)
+
+	// http://ss64.com/nt/icacls.html
+	if err := host.Run("icacls", dir, "/grant:r", user.Name+":(OI)(CI)F"); err != nil {
+		return err
+	}
+
+	setOwnerArgs := []string{dir, "/setowner", user.Name}
+	if recurse {
+		setOwnerArgs = append(setOwnerArgs, "/T")
+	}
+	return host.Run("icacls", setOwnerArgs...)
 }
 
 // The windows implementation of os.Rename(...) doesn't allow renaming files

